client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,576 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use futures::{FutureExt, StreamExt};
+use libp2p::PeerId;
+use log::{error, trace};
+use sc_utils::mpsc::{tracing_unbounded, TracingUnboundedReceiver, TracingUnboundedSender};
+use std::{
+	borrow::Cow,
+	collections::{
+		hash_map::Entry,
+		HashMap, HashSet,
+	},
+	time::{Duration, Instant},
+};
+use wasm_timer::Delay;
```

Comment:
Originally just the `Peerset` legacy, but in our case it seems more convenient because of the [`new_at`](https://docs.rs/wasm-timer/latest/wasm_timer/struct.Delay.html#method.new_at) method (`futures_timer::Delay` only has `new`). And I'm scared of calculating durations with [`Instant::duration_since`](https://doc.rust-lang.org/std/time/struct.Instant.html#method.duration_since) because the doc says it may panic in the future versions instead of saturating.

Of course we can overcome the possible future gotchas of `Instant::duration_since` by adding a little logic if `wasm_timer` is not desirable for some reason.

Reply:
The implementation difficulty here is that if we keep `Delay` instead of the absolute time in `ProtocolController`, we can't `select!` on it, because select takes ownership of the future. This is why I keep absolute time between calls to `next_action()` and construct `Delay` just before selecting on it.

Reply:
I see. We can probably keep it then. It would of course be preferable if the networking code would use only one type of timer but it's probably fine.